zhehedream/Advanced_Dungeons
# A few improvements... (see changelog)

Changelog:
* Add autocompletion to commands
* Change plugin folder name to follow Spigot conventions (see notes)
* Makes the plugin compile without any warnings (see notes)
* Fix: advanceddungeons.enter was permission never tested
* Fix: advanceddungeons.place was permission never tested
* Fix: greymerk.roguelike.worldgen.blocks.Quartz.get() return value
* Fix: greymerk.roguelike.worldgen.WorldEditor.getStat()
  which tested a Block instead of a Material in its stats Map.
* Allows to activate / disable world without restart
* Same for a global activation / disabling (change natural spawn param)
* So, this plugin now support plugman and alternatives
* Colorize plugin messages
* Simplify StringUtils import in ConfigurationProvider.java
* Remove an invisible character in DungeonGenerator.java
* minor bugfixes (like in command you could add not existing worlds)
* A bit of refactoring (also move the main command in its own class)
* Remove unused imports, unnecessary annotations, ...
* Add .gitignore to ignore "target" folder,
  because some devs use maven or IDE which creates it
* (minor impact but can count) remove the test DONATURALSPAWN
  from Dungeon.canSpawnInChunk, because the BlockPopulator is only
  added in this world if this is the case.
* (minor impact but can count) pass values to methods without
  creating variables if not necessary in a few methods,
  like in Dungeon.forceSpawnInChunk() for example.
* Move config from advanced_dungeons.cfg to config.yml to follow
  Spigot convention + will be lighter: Old RogueConfig class
  tested each time you get a value if the config needs to be loaded
  (using the boolean "testing", I suppose this was for dev purposes,
  but now, you can reload the plugin and it reload its values,
  and in fact reproduce what JavaPlugin.getConfig() allows.
  I've also changed code to directly calling values not methods,
  minor perf improvement but as it was called lot of time...
* Also moved enable worlds into config.yml.
* So, removed unused package greymerk.roguelike.config + your class tupple
* removed LOWERLIMIT (lowerLimit in config), as it's never used
  and Config_Detail.txt says "not used".
* Same for SPAWNFREQUENCY / spawnFrequency.
* Same for LOOTING / looting, even if documented, this was not used.
* Same for UPPERLIMIT / upperLimit, not documented, not used.
* removed "doBuiltinSpawn" from the default config,
  as Config_Detail.txt says "must be true" but let SPAWNBUILTIN exists
  with hard coded value "true" in case of needed later (see ADDConfig)
* dungeons location are now stored in a database (sqlite by default)
* Add a command to migrate old log.txt coords into the database
* Add a command to get stats per world
* Add a command to locate the nearest dungeon (+ perm)
* Moved Config_Detail.txt & guide_coexistence_loot.txt info src/docs.
  All files put in it will be cipied into users plugins/AdvancedDungeons/docs.
  (I've then renamed files to have same extension than which one it describes,
  so users can open with same editor and have same syntax coloration).
* Moved lang folder into src to embed them when compiling (required for next point).
* Allows to translate plugin messages
* Better support for I18n: Now, translated strings can be directly injected
  if the users choose a supported language in their config.yml then suppress
  lang.json (or only parts they want to be translated).
* Add a README.md file to illustrate the Github repository

Notes:
------
For users:
* As the plugin folder is renamed to "AdvancedDungeons", user have to
  move their previously created files in it, or simply rename their
  old plugins/advanced_dungeons to plugins/AdvancedDungeons
* They also have to report their settings from advanced_dungeons.cfg
  to config.yml (and advanced_dungeons.cfg can be deleted)
* world.json can also be deleted, worlds have now to be set in config.yml
  into "worlds" parameter (couple <worldname-string>: <boolean-enable>)
* Users can remove lowerLimit, spawnFrequency, doBuiltinSpawn, upperLimit
  and looting (those values will no more been generated in default config).
  If not removed from config file, this has no effect.

For devs:
* To add a new parameter / change default value, you now simply have
  to add / change it into the "config.yml" embed into src folder.
* I don't know (and had not enough time to check deeper) if generate()
  calls are useful in forceGenerateNear() and generateNear()
  of greymerk.roguelike.dungeon.Dungeon class because their results
  were stored in a local variable never used...
  I've suppress the assignation to prevent compile warning,
  but I've let the call, maybe not necessary but in the doubt...
* In zhehe.advanceddungeons.util.BiomeDictionary class:
  private method hasSubTypes() and private class BiomeInfo
  seems to be unused, in the doubt I've let them and only add
  an annotation to suppress compile warning.
* In greymerk.roguelike.treasure.TreasureChest class:
  seed (private variable) is written but never read,
  I've added annotation "@SuppressWarnings("unused")"
* Same for greymerk.roguelike.treasure.loot.WeightedRandomLoot class:
  private field damage never read, so annotation added.
* As the plugin use API 1.14, you can remove "LEGACY_xxx",
  for example in greymerk.roguelike.worldgen.blocks.BlockType.get()
  (I've added @SuppressWarnings("deprecation") to it so)
* Same thing in greymerk.roguelike.treasure.loot.provider.ItemNovelty
  added the annotation for getItem().
* DungeonGenerator.forcePopulate() seems not used, but as it's public,
  this doesn't throw compile warning. Maybe can be removed.
* (could improve perfs if my though is true, I don't know the code enough)
  in DungeonGenerator, new instance of WorldEditor + Dungeon are created
  for each populated chunks.
  - Maybe this could be done once for all in this class constructor?
  - If not, maybe move canSpawnInChunk out of Dungeon not to instanciate
    the full class if the dungeon is not allowed to spawn...
  (I prefer not touch it for now, because don't know if it needs to be
  recalculated each time)
  At least, Dungeon.initResolver() seems to do useless things...
  And this is executed each time an instance is init.
  Maybe I'm wrong but if not, removing all this stuff could be a good idea.
  I didn't took enough time to be sure but a quick look at the code
  let me think this is a remain of the old plugin "Roguelike Dungeons",
  which read the settings dir for kinda addons...
* (details) Globally, there is IMHO a problem with code indentation:
  sometimes it's tabulation, sometimes it's spaces... a bit harder to read
  and this not follow java code conventions (which allows tab or spaces,
  but recommend to be consistent). So, I've tried not touch indents when
  I can, but reflex are hard to calm down (I've shortcuts to re-indent code
  following the project indentations rules... and I cannot set them globally here).
  So, sorry if I made indentation mistakes.
* /!\ Not sure about how you compile... I guess only src content is added
  but when I compile like that, I got a (working) Jar which is less than half size:
  (yours: 1.7Mo, mine 750k - If I include ALL the repository, I've got 1.4Mo)
  So, depending on your method, maybe some file moving into src was not necessary.
